stdlib/public/core/TempAlloc.swift

```diff
@@ -0,0 +1,170 @@
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftShims
+
+/// Will an allocation of the specified size fit on the stack?
+///
+/// - Parameters:
+///   - byteCount: The number of bytes to temporarily allocate. `byteCount` must
+///     not be negative.
+///
+/// - Returns: Whether or not there is sufficient space on the stack to allocate
+///   `byteCount` bytes of memory.
+@_alwaysEmitIntoClient @inline(__always)
+internal func _isStackAllocationSafe(_ byteCount: Int) -> Bool {
+  // Fast path, with a limit known at compile-time. Allocations smaller than
+  // this limit are reasonable to allocate on the stack without worrying about
+  // running out of space, and the compiler would emit such allocations on the
+  // stack anyway when they represent structures or stack-promoted objects.
+  if byteCount <= 1024 {
+    return true
+  }
+
+  // Slower path, with a limit known at runtime and requiring a new standard
+  // library to check.
+  guard #available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *) else {
+    return false
+  }
+  return _swift_stdlib_isStackAllocationSafe(byteCount)
+}
+
+/// Provides scoped access to a raw buffer pointer with the specified byte count
+/// and alignment.
+///
+/// - Parameters:
+///   - byteCount: The number of bytes to temporarily allocate. `byteCount` must
+///     not be negative.
+///   - alignment: The alignment of the new, temporary region of allocated
+///     memory, in bytes.
+///   - body: A closure to invoke and to which the allocated buffer pointer
+///     should be passed.
+///
+/// - Returns: Whatever is returned by `body`.
+///
+/// - Throws: Whatever is thrown by `body`.
+///
+/// This function is useful for cheaply allocating raw storage for a brief
+/// duration. Storage may be allocated on the heap or on the stack, depending on
+/// the required size and alignment.
+///
+/// When `body` is called, the contents of the buffer pointer passed to it are
+/// in an unspecified, uninitialized state. `body` is responsible for
+/// initializing the buffer pointer before it is used _and_ for deinitializing
+/// it before returning, but deallocation is automatic.
+///
+/// The implementation may allocate a larger buffer pointer than is strictly
+/// necessary to contain `byteCount` bytes. The behavior of a program that
+/// attempts to access any such additional storage is undefined.
+///
+/// The buffer pointer passed to `body` (as well as any pointers to elements in
+/// the buffer) must not escape—it will be deallocated when `body` returns and
+/// cannot be used afterward.
+@_alwaysEmitIntoClient @inline(__always)
+public func withUnsafeTemporaryAllocation<R>(byteCount: Int, alignment: Int, _ body: (UnsafeMutableRawBufferPointer) throws -> R) rethrows -> R {
```

Comment:
Formatting nit: we have an 80-character line length limit throughout the compiler & stdlib codebase.

```suggestion
public func withUnsafeTemporaryAllocation<R>(
  byteCount: Int, 
  alignment: Int, 
  _ body: (UnsafeMutableRawBufferPointer) throws -> R
  ) rethrows -> R {
```

(etc for all overlong lines)

Reply:
I'll lint it all when we're sure of names. :) However I see plenty of counter-examples in the stdlib.

Reply:
Mistakes happen, but nevertheless, we have a [hard 80-character line length limit](https://github.com/apple/swift/blob/main/docs/StandardLibraryProgrammersManual.md#formatting-conventions) for the stdlib. We want the stdlib to fit on regulation punch cards. 

More importantly, the 80 character limit makes it possible to read code on GitHub. Adding proper line breaks is just a basic kindness to reviewers -- it helps them understand the changes. 

These function signatures are super critical to this PR, but I currently have to awkwardly scroll around to read them, and even then, they are a pain to understand.

Compare and contrast:

```swift
@_alwaysEmitIntoClient @inline(__always)
public func withUnsafeTemporaryAllocation<R>(byteCount: Int, alignment: Int, _ body: (UnsafeMutableRawBufferPointer) throws -> R) rethrows -> R {
  ...
}
```

vs

```swift
@_alwaysEmitIntoClient @inline(__always)
public func withUnsafeTemporaryAllocation<R>(
  byteCount: Int, 
  alignment: Int, 
  _ body: (UnsafeMutableRawBufferPointer) throws -> R
) rethrows -> R {
  ...
}
```

Presentation matters.

(There is a single pragmatic exception to this rule: overlong precondition message strings are okay to leave as is. This exception most definitely doesn't apply to regular code, much less public API signatures.)

Reply:
I promise I'll clean it up. :)